File: Key-public-cs50-problems-2019-x-caesar/caesar.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>

char print_ciphertext(int x);
string plaintext = "plain";
string ciphertext = "cipher";
int key = 1;
int i = 1;
int main(int argc, string argv[])
   
{
    long digits_check = 1;
    // проверка, что введены только цифры (любое умножение на 0 сделает переменную digits_check равной нулю)
    
    // проверка, что введено только одно число (и ничего больше)
    if (argc == 2)
    {
        for (int j = 0; j < strlen(argv[1]); j++)
        {
            digits_check = digits_check * (isdigit(argv[1][j]) / 2048);       
        }  
        if (digits_check != 0)
        {
            plaintext = get_string("plaintext: ");
        }
        else
        {
            printf("Usage: ./caesar key1\n");
            // завершаем программу, если ввод не соответствует ожидаемому
            return 1;
        }
    }
    else
    {
        printf("Usage: ./caesar key1\n");
        // завершаем программу, если ввод не соответствует ожидаемому
        return 1;
    }
    printf("ciphertext: ");
    ciphertext = plaintext;
    // хитрый способ получать значения 1-26, даже если введено больше
    key = (26 + atoi(argv[1])) % 26;
    int length = strlen(plaintext);
    for (i = 0; i < length; i++)    
    {
        // правила шифрования для букв в нижнем регистре (должны оставаться в нижнем после подмены)
        if (plaintext[i] > 64 && plaintext[i] < 91)
        {
            // хитрая логика, не позволяющая выйти за рамки диапазона
            print_ciphertext(91);
        }
        // правила шифрования для букв в верхнеи регистре (должны оставаться в верхнем после подмены)
        else if (plaintext[i] > 96 && plaintext[i] < 123)  
        {
            print_ciphertext(123);          
        }
        // всё, что не буквы печатается так же, как изначально
        else
        {
            printf("%c", plaintext[i]); 
        }   
    }   
    printf("\n");
    return 0;
}

char print_ciphertext(int x)
// хитрая логика, не позволяющая выйти за рамки диапазона
{
    if ((int) plaintext[i] + key >= x) // х == 91 или 123
    {
        ciphertext[i] = plaintext[i] - 26 + key;
        printf("%c", ciphertext[i]);
        return ciphertext[i];
    }
    else
    {
        ciphertext[i] = plaintext[i] + key;
        printf("%c", ciphertext[i]);
        return ciphertext[i];
    }
}

File: source/pset6/caesar/caesar.py
from cs50 import get_string
from sys import argv, exit


def main():
    # проверка, что введено только одно число (и ничего больше)
    if len(argv) == 2 and argv[1].isdigit() and int(argv[1]) >= 0:
        plaintext = get_string("plaintext: ")
    else:
        print("Usage: python caesar.py key")
        exit(1)

    # сохраняем ключ, предварительно поделив его по модулю на 26 (чтобы учесть ввод > 26)
    key = int(argv[1]) % 26

    # заголовок для будущего вывода
    print("ciphertext: ", end="")

    # начинаем шифровку
    for c in plaintext:
        if ord(c) > 64 and ord(c) < 91:
            print_ciphertext(91, c, key)
        elif ord(c) > 96 and ord(c) < 123:
            print_ciphertext(123, c, key)
        else:
            print(c, end="")
    print()
    exit(0)


def print_ciphertext(x, c, key):
    if (ord(c) + key) >= x:
        print(chr(ord(c) - 26 + key), end="")
    else:
        print(chr(ord(c) + key), end="")


main()
